File: nips/tf_ops/points_build_assignment/points_build_assignment.py
import numpy as np
import tensorflow as tf
from tensorflow.python.framework import ops
import os
import sys
BASE_DIR = os.path.dirname(os.path.abspath(__file__))
sys.path.append(BASE_DIR)
points_build_assignment_module = tf.load_op_library(os.path.join(BASE_DIR, 'points_build_assignment_so.so'))

def points_build_assignment(x, y):
  """
  Args
    x: BxMxN float tensor of point log probabilities
    y: BxM int tensor of point class labels

  Returns
    cost_matrix: BxNxN float tensor
  """
  return points_build_assignment_module.points_build_assignment(x, y)
@tf.RegisterShape('PointsBuildAssignment')
def _points_build_assignment_shape(op):
  shape1 = op.inputs[0].get_shape().with_rank(3)
  shape2 = op.inputs[1].get_shape().with_rank(2)
  return [tf.TensorShape([shape1.dims[0], shape1.dims[2], shape1.dims[2]])]

def my_func(x, y, grad_cost_matrix): # grad_cost_matrix is the grad of the output
  b,m,n = x.shape
  grad = np.zeros_like(x).astype('float32')
  for bidx in range(b):
    for i in range(m):
        for k in range(n):
          grad[bidx,i,k] = grad_cost_matrix[bidx, k, y[bidx,i]]
  return grad

@tf.RegisterGradient('PointsBuildAssignment')
def _points_build_assignment_grad(op, grad_cost_matrix): # grad is BxNxN
  x = op.inputs[0] # BxMxN
  y = op.inputs[1] # BxM
  # {grad_x}_i_k) = {grad_cost_matrix}_k_yi 
  return [tf.py_func(my_func, [x,y,grad_cost_matrix], tf.float32), None]


if __name__=='__main__':
  with tf.Graph().as_default():
    with tf.device('/cpu:'+str(0)):
      x = tf.Variable(np.random.randn(16,2,3).astype('float32'))
      y = tf.Variable(np.random.randint(low=0, high=3, size=(16,2)).astype('int32'), dtype=tf.int32)
      cost_matrix = points_build_assignment(x, y)
      print cost_matrix
    with tf.Session('') as sess:
      sess.run(tf.initialize_all_variables())
      x_val, y_val, cost_matrix_val = sess.run([x, y, cost_matrix])
      print x_val[0,...].squeeze()
      print y_val[0,...].squeeze()
      print cost_matrix_val[0,...].squeeze()

File: nips/tf_ops/points_build_assignment/points_build_assignment.cc
#include <string.h>
#include "tensorflow/core/framework/op.h"
#include "tensorflow/core/framework/shape_inference.h"
#include "tensorflow/core/framework/op_kernel.h"

using namespace tensorflow;

/*
Build an assignment cost matrix from image pixel probabilities.

x: BxMxN, 3-D probability map of a point cloud for per-point class prediction.
y: BxM, 2-D label for each point.
cost_matrix: BxNxN, 3-D cost matrix for assignment matrix,
  cost_matrix[:,i,j] is the score that pred-class-i matches pred-class-j
*/
REGISTER_OP("PointsBuildAssignment")
    .Input("x: float32")
    .Input("y: int32")
    .Output("cost_matrix: float32");
REGISTER_OP("PointsBuildAssignmentGrad")
    .Input("x: float32")
    .Input("y: int32")
    .Output("grad: float32"); // grad is for x

void build_assignment_matrix(int b, int m, int n, const float* x, const int* y, float* cost_matrix) {
	for (int k=0;k<b;k++){
		for (int i=0;i<m;i++){
			for (int l=0;l<n;l++){
				cost_matrix[l*n+y[i]] += x[i*n+l];
			}
		}
		x += m*n;
		y += m;
		cost_matrix += n*n;
	}
}

class PointsBuildAssignmentOp : public OpKernel {
	public:
		explicit PointsBuildAssignmentOp(OpKernelConstruction* context) : OpKernel(context) {}

		void Compute(OpKernelContext* context) override {
			const Tensor& x_tensor = context->input(0);
			OP_REQUIRES(context, x_tensor.dims() == 3, errors::InvalidArgument("input must be 3-D BxMxN."));
			auto x_flat = x_tensor.flat<float>();
			const float* x = &(x_flat(0));
			int b = x_tensor.shape().dim_size(0);
			int m = x_tensor.shape().dim_size(1);
			int n = x_tensor.shape().dim_size(2);

			const Tensor& y_tensor = context->input(1);
			OP_REQUIRES(context, y_tensor.dims() == 2, errors::InvalidArgument("input must be 2-D BxM."));
			auto y_flat = y_tensor.flat<int>();
			const int* y = &(y_flat(0));

			Tensor* cost_matrix_tensor = NULL;
			OP_REQUIRES_OK(context, context->allocate_output(0, TensorShape{b,n,n}, &cost_matrix_tensor)); 
			auto cost_matrix_flat = cost_matrix_tensor->flat<float>();
			float* cost_matrix = &(cost_matrix_flat(0));
         		memset(cost_matrix, 0, sizeof(float)*b*n*n);

			build_assignment_matrix(b,m,n,x,y,cost_matrix);
		}
};

REGISTER_KERNEL_BUILDER(Name("PointsBuildAssignment").Device(DEVICE_CPU), PointsBuildAssignmentOp);
